### DISCUSSION TITLE
mfsmount cpu consumption bottleneck

### DISCUSSION THREAD

**Discussion by Yaaaqi** [moosefs/moosefs#463] [Q&A]
Hi everyone,

I do get some varying performance numbers from different clients I'm testing out moose cluster on.
They range from over 10Gbit/s to sub 2Gbit/s, even on the same client. It hugely depends on the application used to draw data.
Generally, my observation is that everything that is pulling different files/parts of files in parallel is performing very well and tools that are pulling strictly underperforme. Especially visible has this been on LTO backup tools for example which seem not to be optimized for network access.

Following this observation, I tried to find the bottleneck by using rclone copy to simulate single file access and the numbers seem to line up.
Checking the processes mfsmount seemed to run at about 100% which indicates that this is not multithreaded and caps performance.

   rclone copy -P --transfers 1 /mnt/mfs/src.mov /mnt/mfs/target.mov

First question: Am I right with these observations and conclusions or am I missing something?
Second question: Are there any configurations/tricks to get around this limitation or optimize it?

Already thinking about using a massively singlecore optimized cpu as smb/nfs gateway. Unfortunately, my tests got me only to about 4Gbit/s with a 5GHz cpu which makes me doubt that there is a reasonable solution for going to 6Gbit/s+. I was also wondering if macos m1 rosetta2 abstraction could do an abstraction that results in distrubuted compute of the process.

thanks for any feedback and ideas,
Jacob

![image](https://user-images.githubusercontent.com/49562362/150774994-870f2325-29b3-41d7-a46f-e89ba9128b38.png)

**Comment by xandrus** [comment]
Hi,
I believe that we have more topics in this one thread here:
1.) Coping files inside one MooseFS mount.
2.) MooseFS mount and threads?
3.) Using rclone software
4.) Re-sharing MooseFS resources with NFS/CIFS gates

So:
1.) If you like to make a copy file inside MooseFS cluster please consider using mfsmakesnapshot command. This is extremely fast operation made on metadata level - so you will not use network to transfer files.

2.) MooseFS mount is multi thread application.
Please check if rclone will use more threads in case of copping only one file.

3.) You have to check what block size is used during rclone operation. Please check:
```cat /mnt/mfs/.oplog | tee ~/mfsop.log```
You should see something like this
```
01.27 08:33:56.658307: uid:0 gid:0 pid:382165 cmd:read (110525,131072,267911168) [handle:01000001]: OK (131072)
01.27 08:33:56.658355: uid:0 gid:0 pid:382165 cmd:read (110525,131072,268042240) [handle:01000001]: OK (131072)
01.27 08:33:56.658392: uid:0 gid:0 pid:382165 cmd:read (110525,131072,268173312) [handle:01000001]: OK (131072)
```
After OK you should see current read operation block size

4.) Can you tell us something more abut this configuration?
What is you network latency between MFSmaster - client and between NFS/CIFS gate and end user.
Please keep in mind that SAMBA is single-thread - I mean single thread for each connection.

**Reply by Yaaaqi** [reply]
Hi Xandrus,

thanks a lot for your reply.
to 1) copying is not actually what I'm looking for, but good to know obout the snapshot command. Will certainly be usefule some time in the future.
to 2) Thanks for clearifying. More below.
Seems to me rclone is multithreading, at least the processes indicate this. while screenshot I was running with --transfers=1.
![Screenshot 2022-02-01 at 23 29 15](https://user-images.githubusercontent.com/49562362/152062469-dbde606d-1bd6-4d58-98a1-476ee341c845.png)

to 3) Rclone is actually just a testvehicle to approx the behaviour of Archiware (LTO) and Davinci Resolve (Playback/render media files) which are the two tools I try to optimize. Rclone itself seems to use 1048576 for read and 32768 for write.

```
02.01 21:38:59.200925: uid:1000 gid:1000 pid:661566 cmd:write (7741088,32768,3141890048) [handle:01000002]: OK (32768)
02.01 22:07:57.908600: uid:0 gid:0 pid:30215 cmd:read (7739265,1048576,857731072) [handle:02000001]: OK (1048576)
```

I'll try to get further checks about the blocksize with the software in question and will share them.
Would there be any way to optimize for different blocksizes?

to 4) 
Ping Latencies are between 0.2-0.6ms for all connections between nodes, master, mfsclients, smb-gateway and smb-client.
All interfaces 10 or 100 gbit/s. 

I set up the smb server to use multipath so it should open multiple connections for a request, though I have to admit I'm not exactly sure if that's working out fine. Haven't dived deep into this end. Basically smb again is rather a try to alter the access behaviour to improve performance.

__________

To give a less abstract and more transparent picture of what looked strange to me:

Davinci Resolve Playback of a 
  - Imagesequence is steadily running fine with a approx 8gbit/s test sequence
  - Quicktime also with aim at about 8gbit/s is only running at about3 fps and effective network bandwith of 1,5gbit/s
  - files are all stored on the same folder and mfsnode with same mfs storage class
  - Mac workstation via mfsclient mount

Daylight (as comparison to resolve) Playback of a 
  - Imagesequence is not running fluidly, capped at approx 5gbit/s test sequence
  - Quicktime is steadily running fine with a approx 8gbit/s test sequence
  - same files as above
  - CentOS8 workstation via mfsclient mount

Archiware LTO-Job
  - running only at about 0,8gbit/s archiving large quicktime files
  - rclone with only one parallel copythread running at about same speed
  - rclone with 10 parallel copythreads running up to about 7gbit/s
  - ubuntu vm via mfsclient mount

Sidenotes: 
  - backchecking these scenarios from local ssds gives full performance in all cases.
  - Rclone throughput was scaling pretty linear with parallel threads on all above clients

I know that there are various factors altering for all of those test scenarios, so it's hard to draw a proper conclusion. Overall my impression is that it has something to do with how the tools are accessing the files. 
Max throughput never seems to be an issue. 
RAM and CPU all look relaxed. 
Hence my thought that maybe software that is using multiple threads at once for accessing files (preread on image sequences in resolve's case and even for single quicktime files in case of daylight) is doing well and strictly linear/single thread access could be bottlenecked. 
Another indication for this seemd to me that rclone throughput with 1 parallel copy was roughly scaling with client single core cpu power and one mfsmount process was usually running at about 100% cpu in htop/top and all other mfsmount processes were very low. 
Maybe I misunderstood how mfsmount is multithreading load. 
Will mfsmount multithread by putting load to one mfsmount process after the other, as soon as one of them reaches about 100% cpu, assuming it gets hit by a non parallelized file read? This would perfectly fit to what I monitored.

**Reply by Yaaaqi** [reply]
Resolve on mac and baselight on mac are both using 65k blocks
```

03.11 16:35:42.826145: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575597568) [handle:17000003]: OK (65536)
03.11 16:35:42.826182: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575663104) [handle:17000003]: OK (65536)
03.11 16:35:42.826224: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575728640) [handle:17000003]: OK (65536)
03.11 16:35:42.826262: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575794176) [handle:17000003]: OK (65536)
03.11 16:35:42.826300: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575859712) [handle:17000003]: OK (65536)
03.11 16:35:42.826335: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575925248) [handle:17000003]: OK (65536)
03.11 16:35:42.826392: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66575990784) [handle:17000003]: OK (65536)
03.11 16:35:42.826444: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66576056320) [handle:17000003]: OK (65536)
03.11 16:35:42.826492: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66576121856) [handle:17000003]: OK (65536)
03.11 16:35:42.826538: uid:506 gid:20 pid:75583 cmd:read (7741085,65536,66576187392) [handle:17000003]: OK (65536)
03.11 16:35:42.826571: uid:506 gid:20 pid:75583 cmd:read (7741085,32788,66576252928) [handle:17000003]: OK (32788)
03.11 16:35:42.826604: uid:506 gid:20 pid:56264 cmd:access (7741085,0x0): OK
03.11 16:35:42.826630: uid:506 gid:20 pid:56264 cmd:access (7741085,0x0): OK
03.11 16:35:42.826666: uid:506 gid:20 pid:56264 cmd:getattr (7741085) [no handle]: OK (0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.826813: uid:506 gid:20 pid:56264 cmd:getattr (7741085) [no handle]: OK (0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.827923: uid:506 gid:20 pid:56264 cmd:lookup (1,1ssd_class): OK (1.0,46820,0.0,[drwxrwxrwx:0040777,12,0,0,1647012489,1646998805,1646998805,3050607])
03.11 16:35:42.828037: uid:506 gid:20 pid:56264 cmd:lookup (46820,uncompressed_mov 1): OK (1.0,31301,0.0,[drwxrwxr-x:0040775,2,0,0,1646998098,1644437979,1644437979,3017807])
03.11 16:35:42.828051: uid:506 gid:20 pid:56264 cmd:lookup (1,1ssd_class): OK (1.0,46820,0.0,[drwxrwxrwx:0040777,12,0,0,1647012489,1646998805,1646998805,3050607])
03.11 16:35:42.828179: uid:506 gid:20 pid:56264 cmd:lookup (31301,ddg.mov): OK (0.0,7741085,0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.828186: uid:506 gid:20 pid:56264 cmd:lookup (46820,uncompressed_mov 1): OK (1.0,31301,0.0,[drwxrwxr-x:0040775,2,0,0,1646998098,1644437979,1644437979,3017807])
03.11 16:35:42.828225: uid:506 gid:20 pid:56264 cmd:access (7741085,0x0): OK
03.11 16:35:42.828254: uid:506 gid:20 pid:56264 cmd:getattr (7741085) [no handle]: OK (0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.828343: uid:506 gid:20 pid:56264 cmd:lookup (31301,ddg.mov): OK (0.0,7741085,0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.828372: uid:506 gid:20 pid:56264 cmd:access (7741085,0x0): OK
03.11 16:35:42.828391: uid:506 gid:20 pid:56264 cmd:getattr (7741085) [no handle]: OK (0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.829107: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576252928) [handle:17000003]: OK (65536)
03.11 16:35:42.829164: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576318464) [handle:17000003]: OK (65536)
03.11 16:35:42.829208: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576384000) [handle:17000003]: OK (65536)
03.11 16:35:42.829253: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576449536) [handle:17000003]: OK (65536)
03.11 16:35:42.829321: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576515072) [handle:17000003]: OK (65536)
03.11 16:35:42.829365: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576580608) [handle:17000003]: OK (65536)
03.11 16:35:42.829417: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576646144) [handle:17000003]: OK (65536)
03.11 16:35:42.829499: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576711680) [handle:17000003]: OK (65536)
03.11 16:35:42.829548: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576777216) [handle:17000003]: OK (65536)
03.11 16:35:42.829593: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576842752) [handle:17000003]: OK (65536)
03.11 16:35:42.829642: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576908288) [handle:17000003]: OK (65536)
03.11 16:35:42.829708: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66576973824) [handle:17000003]: OK (65536)
03.11 16:35:42.829755: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577039360) [handle:17000003]: OK (65536)
03.11 16:35:42.829805: uid:506 gid:20 pid:56264 cmd:lookup (1,1ssd_class): OK (1.0,46820,0.0,[drwxrwxrwx:0040777,12,0,0,1647012489,1646998805,1646998805,3050607])
03.11 16:35:42.829807: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577104896) [handle:17000003]: OK (65536)
03.11 16:35:42.829876: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577170432) [handle:17000003]: OK (65536)
03.11 16:35:42.829915: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577235968) [handle:17000003]: OK (65536)
03.11 16:35:42.829949: uid:506 gid:20 pid:56264 cmd:lookup (46820,uncompressed_mov 1): OK (1.0,31301,0.0,[drwxrwxr-x:0040775,2,0,0,1646998098,1644437979,1644437979,3017807])
03.11 16:35:42.829970: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577301504) [handle:17000003]: OK (65536)
03.11 16:35:42.830013: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577367040) [handle:17000003]: OK (65536)
03.11 16:35:42.830060: uid:506 gid:20 pid:56264 cmd:lookup (31301,ddg.mov): OK (0.0,7741085,0.0,[-rw-rw-r--:0100664,1,506,20,1647012907,1636456137,1642803196,191201321300])
03.11 16:35:42.830068: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577432576) [handle:17000003]: OK (65536)
03.11 16:35:42.830119: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577498112) [handle:17000003]: OK (65536)
03.11 16:35:42.830173: uid:506 gid:20 pid:75581 cmd:read (7741085,65536,66577563648) [handle:17000003]: OK (65536)

```

**Reply by Yaaaqi** [reply]
daylight on linux is using 128k blocks

`03.11 17:03:15.809014: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4072479768) [handle:75000023]: OK (131072)
03.11 17:03:15.809627: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4072610840) [handle:75000023]: OK (131072)
03.11 17:03:15.810433: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4072741912) [handle:75000023]: OK (131072)
03.11 17:03:15.811025: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4072872984) [handle:75000023]: OK (131072)
03.11 17:03:15.811751: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073004056) [handle:75000023]: OK (131072)
03.11 17:03:15.812384: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073135128) [handle:75000023]: OK (131072)
03.11 17:03:15.813214: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073266200) [handle:75000023]: OK (131072)
03.11 17:03:15.813792: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073397272) [handle:75000023]: OK (131072)
03.11 17:03:15.814458: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073528344) [handle:75000023]: OK (131072)
03.11 17:03:15.815039: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073659416) [handle:75000023]: OK (131072)
03.11 17:03:15.815728: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073790488) [handle:75000023]: OK (131072)
03.11 17:03:15.816289: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4073921560) [handle:75000023]: OK (131072)
03.11 17:03:15.818242: uid:506 gid:506 pid:25859 cmd:read (78517,131072,4074052632) [handle:75000023]: OK (131072)`
